subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java

```diff
@@ -54,27 +73,63 @@ public DefaultJvmTestSuite(String name, ConfigurationContainer configurations, D
-                if (framework instanceof JUnitPlatformTestingFramework) {
-                    return "org.junit.jupiter:junit-jupiter-engine:" + framework.getVersion().get();
-                } else {
-                    return getObjectFactory().fileCollection();
-                }
-            }));
+        this.attachedDependencies = false;
+        // This complexity is to keep the built-in test suite from automatically adding dependencies
+        // unless a user explicitly calls one of the useXXX methods
+        // Eventually, we should deprecate this behavior and provide a way for users to opt out
+        // We could then always add these dependencies.
+        this.attachDependencyAction = x -> attachDependenciesForTestFramework(dependencies, implementation);
+
+        if (!name.equals(JvmTestSuitePlugin.DEFAULT_TEST_SUITE_NAME)) {
+            useJUnitJupiter();
+        } else {
+            // for the built-in test suite, we don't express an opinion, so we will assume JUnit4 and not add any dependencies
+            // if a user explicitly calls useJUnit or useJUnitJupiter, the built-in test suite will behave like a custom one
+            // and add dependencies automatically.
+            getTestingFramework().convention(new TestingFramework(Frameworks.NONE, null));
         }
 
         this.targets = getObjectFactory().polymorphicDomainObjectContainer(JvmTestSuiteTarget.class);
         this.targets.registerBinding(JvmTestSuiteTarget.class, DefaultJvmTestSuiteTarget.class);
 
         this.dependencies = getObjectFactory().newInstance(DefaultComponentDependencies.class, implementation, compileOnly, runtimeOnly);
+
+        addDefaultTestTarget();
+
+        this.targets.withType(JvmTestSuiteTarget.class).configureEach(target -> {
+            target.getTestTask().configure(task -> {
+                task.getTestFrameworkProperty().convention(getTestingFramework().map(framework -> {
+                    switch(framework.framework) {
+                        case NONE:
+                        case JUNIT4:
+                            return new JUnitTestFramework(task, (DefaultTestFilter) task.getFilter());
+                        case JUNIT_JUPITER:
+                            return new JUnitPlatformTestFramework((DefaultTestFilter) task.getFilter());
+                        default:
+                            throw new IllegalStateException("do not know how to handle " + framework);
+                    }
+                }));
+            });
+        });
+
+    }
+
+    private void attachDependenciesForTestFramework(DependencyHandler dependencies, Configuration implementation) {
+        if (!attachedDependencies) {
+            dependencies.addProvider(implementation.getName(), getTestingFramework().map(framework -> {
+                switch (framework.framework) {
+                    case JUNIT4:
+                        assert framework.version != null;
```

Comment:
Should Java `assert`s remain?

Reply:
I think this is OK. We have these sprinkled in other places, but we don't run Gradle with assertions enabled.

Reply:
I'm less concerned with the overhead of checking, but more surprised Gradle doesn't have a Checkstyle rule or something in place to ban these, actually.  

I always subscribed to the idea that if something is worth checking in production code, then that check shouldn't be disable-able - like Java assertions, and if it isn't worth checking, why leave an optional check to make it less readable?  

In a case like this, which is input validation-ish, would Guava `Preconditions.checkArgument()` be a better choice?

Reply:
I'm with Tom - remove `assert` statements from production Java code. If this is important enough to need a null check, use a try/catch block or `Objects#requireNonNull`

Reply:
all right, I can rearrange this a bit